### PR TITLE
Add support to ShellExecute in powershell core when it's running on Windows full SKUs

### PR DESCRIPTION
**Pull request by daxian-dbw**
Add support to ShellExecute in powershell core when it's running on Windows full SKUs.
Updated `Get-Help -Online`, `Invoke-Item` and `Start-Process` to support start a process using ShellExecute on powershell core on windows full SKUs.

TODO: In .NET Core, managed threads are all eagerly initialized with MTA mode, so to call `ShellExecuteEx` from a STA thread, we need to create a native thread using `CreateThread` function and initialize COM with STA on that thread. In #3281, we are calling `ShellExecuteEx` directly on MTA thread, and it works for things like opening a folder in File Explorer, opening a file with the application that is associated with its extension in shell, opening URL in web browser and etc, but it's not guaranteed to work in all ShellExecution scenarios. #2969 will be used to track the "invoke-on-STA-thread" work.

### CODE REVIEW

File: src/System.Management.Automation/engine/Utils.cs [hunk 1]
```diff
@@ -1588,4 +1595,214 @@ public static void SetTestHook(string property, bool value)
...
+            }
+            finally
+            {                
+                if (shellExecuteInfo.lpFile != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpFile);
+                if (shellExecuteInfo.lpVerb != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpVerb);
+                if (shellExecuteInfo.lpParameters != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpParameters);
+                if (shellExecuteInfo.lpDirectory != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpDirectory);
+            }
+
+            Process processToReturn = null;
```
Comment (Francisco-Gamino) [inline src/System.Management.Automation/engine/Utils.cs]: Could you please write a comment on what is happening right here? Please correct me if I am wrong. In the finally block you are freeing memory for lpfile, lpVerb, lpParameters, and lpDirectory. Then you use the pointer to the process to fetch the process using its id. My question is: Do we need to free anymore memory from shellExecuteInfo once the process is returned?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 2]
```diff
@@ -1330,7 +1330,19 @@ protected override void InvokeDefaultAction(string path)
...
+                    // Try Process.Start first. This works for executables even on headless SKUs.
+                    invokeProcess.StartInfo.FileName = path;
+                    invokeProcess.Start();
+                }
+                catch (Win32Exception)
+                {
+                    // If it's headless SKUs, rethrow.
+                    if (Platform.IsNanoServer || Platform.IsIoT) { throw; }
+                    // If it's full Windows, then try ShellExecute.
+                    ShellExecuteHelper.Start(invokeProcess.StartInfo);
```
Comment (Francisco-Gamino) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Is there a possibility that this call will throw? If it does, how are you handling it?

**Comment by Francisco-Gamino** [comment]
Hi @daxian-dbw, I left some minor comments/questions. Overall it looks very good. Thanks for enabling this!

File: src/System.Management.Automation/engine/Utils.cs [hunk 3]
```diff
@@ -1588,4 +1595,214 @@ public static void SetTestHook(string property, bool value)
...
+        /// <remarks>
+        /// Quoted from MSDN:
+        /// "Because ShellExecuteEx can delegate execution to Shell extensions (data sources, context menu handlers, verb implementations) 
+        ///  that are activated using Component Object Model (COM), COM should be initialized before ShellExecuteEx is called. Some Shell
+        ///  extensions require the COM single-threaded apartment (STA) type. In that case, COM should be initialized as shown here:
+        ///     CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE)
+        ///  There are instances where ShellExecuteEx does not use one of these types of Shell extension and those instances would not require
+        ///  COM to be initialized at all. Nonetheless, it is good practice to always initalize COM before using this function."
+        ///
+        /// TODO: In .NET Core, managed threads are all eagerly initialized with MTA mode, so to call 'ShellExecuteEx' from a STA thread, we
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Utils.cs]: > #2969 will be used to track the "invoke-on-STA-thread" work.

Yes, it's recorded in the description of this PR. This `Start` method will be touched when fixing that issue and this `TODO` can be removed then.

File: src/System.Management.Automation/engine/Utils.cs (hunk 1 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Utils.cs]: No more resource needs to be freed. This code mainly comes from [.NET code here](https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/Process.cs,40054df689616fb5)

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 4]
```diff
@@ -1330,7 +1330,19 @@ protected override void InvokeDefaultAction(string path)
                 invokeProcess.StartInfo.Arguments = path;
                 invokeProcess.Start();
 #elif CORECLR
-                throw new PlatformNotSupportedException();
+                try
+                {
+                    // Try Process.Start first. This works for executables even on headless SKUs.
+                    invokeProcess.StartInfo.FileName = path;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Yes I did. It's able to start executable files, like .exe file, on NanoServer.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 2 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: If it throws, then the exception just propagate and powershell will handle it. It's like what would happen when `invokeProcess.Start()` throws.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 5]
```diff
@@ -1,86 +1,120 @@
...
-            Start-Process -Verb runas -FilePath $pingCommand -ArgumentList $pingParam
-            throw "No Exception!"
-        }
-        catch
-        {
-            $_.FullyQualifiedErrorId | Should be "NotSupportedException,Microsoft.PowerShell.Commands.StartProcessCommand"
-            $_.Exception.Message | Should match '-Verb'
-        }
+    ## -Verb is supported in PowerShell core on Windows full desktop.
+    It "Should give an error when -Verb parameter is used" -Skip:$isFullWin {
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: No, `-Verb` is not supported on non-windows platforms and windows headless SKUs. So on NanoServer and IoT, using `-Verb` should result in a terminating error.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 6]
```diff
@@ -1,86 +1,120 @@
...
+        { Start-Process -Verb runas -FilePath $pingCommand } | ShouldBeErrorId "NotSupportedException,Microsoft.PowerShell.Commands.StartProcessCommand"
+    }
+
+    ## -WindowStyle is supported in PowerShell core on Windows full desktop.
+    It "Should give an error when -WindowStyle parameter is used" -Skip:$isFullWin {
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Same reason as above.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1 [hunk 7]
```diff
@@ -1,86 +1,120 @@
...
+        Test-Path $fileToWrite | Should Be $true
+    }
+
+    It "Should start notepad.exe with ShellExecute" -Skip:(!$isFullWin) {
+        $process = Start-Process notepad -PassThru -WindowStyle Normal
+        $process.Name | Should Be "notepad"
+        $process | Stop-Process
+    }
+
+    It "Should open the application that associates with extension '.txt'" {
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1]: Good catch. I will fix it.

File: src/System.Management.Automation/engine/Utils.cs (hunk 3 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Utils.cs]: I will add the issue number in the comment here as well.

**Comment by daxian-dbw** [comment]
@Francisco-Gamino I have addressed your comments. Could you please take another look?

File: src/System.Management.Automation/engine/Utils.cs (hunk 1 above)
Comment (Francisco-Gamino) [inline src/System.Management.Automation/engine/Utils.cs]: Ok. Thanks.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 4 above)
Comment (Francisco-Gamino) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Awesome. Thanks!

**Comment by Francisco-Gamino** [comment]
@daxian-dbw: LGTM, thanks for getting this done!
ship-it-squirrel

**Pull request merged**